sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -48,8 +48,18 @@ object CostBasedJoinReorder extends Rule[LogicalPlan] with PredicateHelper {
           if projectList.forall(_.isInstanceOf[Attribute]) =>
           reorder(p, p.output)
       }
-      // After reordering is finished, convert OrderedJoin back to Join
+
+      // Cleanups
       result transformDown {
+        // if a Project was created to keep output attribute order after join reordering, but
```

Comment:
is this an improvement? Maybe we should do it in another PR, so that this PR can be backported to old branches.

Reply:
This isn't really an improvement. It's here mainly to help pass existing tests -- I'm adding new projections to fix the output attribute order problem, but these extra projections in the middle would make (`expected`) query plans look pretty ugly.

So consider this is two things that can cancel each other out:
1. Add projections to fix output attribute order;
2. If one of these extra projections is in the middle, get rid of it.

The cleanup (2) is only meant to clean extra projections created because of (1). Both (1) and (2) are in this PR so I don't consider this as a performance improvement over existing stuff.